rfd/0178-github-proxy.md

```diff
@@ -0,0 +1,853 @@
+
+First, she configures a new OAuth app on GitHub side, with the following info:
+- Application Name: <teleport-cluster-name>
+- Homepage URL: <teleport-proxy-addr>
+- Authorization callback URL:: <teleport-proxy-addr>/v1/webapi/github/callback
+
+Then she creates a GitHub integration using `tctl create`:
+```yaml
+kind: integration
+sub_kind: github
+version: v1
+metadata:
+  name: github-my-org
+spec:
+  github:
+    organization: my-org
+    proxy:
+      connector:
+        client_id: <oauth-app-client-id>
+        client_secret: <oauth-app-client-secret>
+        redirect_url: <proxy-address>/v1/webapi/github/callback
+```
+
+Next she exports the SSH CA and imports it to her GitHub organization:
+```shell
+$ tctl auth export --integration github-my-org --type github
+ssh-rsa <cert-pem...>
+
+Go to https://github.com/organizations/my-org/settings/security and click "New
+CA" in the "SSH certificate authorities" section.
+
+Copy-paste the above certificate there and click "Add CA". The CA should have
+the following sha256 fingerprint:
+<fingerprint...>
+```
+
+Next, she creates a GitHub proxy server for Teleport users to access, using
+`tctl create`:
+```yaml
+kind: git_server
+sub_kind: github
+spec:
+  github:
+    integration: github-my-org
+    organization: my-org
+version: v2
+```
+
+To provide access, Alice creates the following role and attach it to desired
+Teleport users:
+```yaml
+kind: role
+metadata:
+  name: github-my-org-access
+spec:
+  allow:
+    github_permissions:
+    - orgs:
+      - my-org
+ ...
+version: v7
+```
+
+Alternatively, Alice can map `{{internal.github_orgs}}` or
+`{{external.github_orgs}}` traits of the users instead of hardcoding the org in
+the role.
+
+#### Bob clones a new Git repository
+
+Bob, a Teleport user that's granted access to the GitHub organization, wants to
+clone a repo.
+
+He logins the `tsh`, and run `tsh git ls` to see what access he has:
+```shell
+$ tsh git ls
+Type   Organization Username  URL                                     
+------ ------------ --------- -------------------------
+GitHub my-org       (n/a)*    https://github.com/my-org
+
+(n/a)*: Usernames will be retrieved automatically upon git commands.
+        Alternatively, run `tsh git login --github-org <org>`.
+
+hint: use 'tsh git clone <git-clone-ssh-url>' to clone a new repository
+      use 'tsh git config update' to configure an existing repository to use Teleport
+      once the repository is cloned or configured, use 'git' as normal
+```
+
+He goes on `github.com` and finds the SSH clone url for the repository he wants
+to clone. Then he runs the `tsh git clone` command with copied url:
+```shell
+$ tsh git clone git@github.com:my-org/my-repo.git
+```
+
+The first `git` command (including the `clone`) will open a browser window to
+trigger the GitHub OAuth flow for Teleport to grab Bob's GitHub ID and
+username. Once Bob sees "Login Successful" from the brower and goes back to his
+terminal.
```

Comment:
What would the flow be in a non-interactive/headless scenario? I.e. if you want to do this from a development box where there is no browser.